pdns/dnssecinfra.cc

```diff
@@ -492,16 +492,34 @@ string hashQNameWithSalt(const NSEC3PARAMRecordContent& ns3prc, const DNSName& q
 string hashQNameWithSalt(const std::string& salt, unsigned int iterations, const DNSName& qname)
 {
   unsigned int times = iterations;
-  unsigned char hash[20];
-  string toHash(qname.toDNSStringLC());
-
-  for(;;) {
-    toHash.append(salt);
-    SHA1((unsigned char*)toHash.c_str(), toHash.length(), hash);
-    toHash.assign((char*)hash, sizeof(hash));
-    if(!times--)
+  unsigned char hash[SHA_DIGEST_LENGTH];
+  string toHash(qname.toDNSStringLC() + salt);
+  if (toHash.capacity() < (salt.size() + sizeof(hash))) {
+    toHash.reserve(salt.size() + sizeof(hash));
```

Comment:
I'm wondering now if that call to `std::string::reserve()` is not a misoptimization:
- when `iterations` == 1, which seems to happen quite often, we might be just doing a needless reallocation if the string is already large enough to hold the hash ;
- with a larger number of iterations, doing the reallocation now or after the first iteration does not really matter.

I'll remove these lines.